src/iocost/iocost.c

```diff
@@ -0,0 +1,299 @@
+               "\nSee the %s for details.\n",
+               program_invocation_short_name,
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+enum HwDbParseState {
+        MODEL = 1,
+        QOS = 2,
+        SKIP = 3,
+        DONE = 4,
+};
+
+static int hwdb_query_for_path(const char *path, sd_hwdb **hwdb, char **modalias) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        const char *model_name;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_hwdb_new(hwdb);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open hwdb: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0)) {
+                log_info("Model name for device %s is unknown", path);
+                return 0;
+        }
+
+        if (r < 0)
+                return log_error_errno(r, "Model name for device %s is unknown", path);
+
+        asprintf(modalias, "block:devname:%s:name:%s", path, model_name);
+
+        return 0;
+}
+
+static char *name_from_key(const char *key) {
+        _cleanup_strv_free_ char **key_parts = NULL;
+
+        key_parts = strv_split(key, "_");
+        return ascii_strlower(strdup(key_parts[2]));
+}
+
+static int apply_solution_for_path(const char *path, const char *name_to_apply) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *modalias = NULL;
+        _cleanup_(sd_hwdb_unrefp) sd_hwdb *hwdb = NULL;
+        enum HwDbParseState state = MODEL;
+        _cleanup_free_ char *name = NULL;
+        _cleanup_free_ char *qos = NULL;
+        _cleanup_free_ char *model = NULL;
+        const char *key, *value;
+        dev_t dev;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_device_get_devnum(device, &dev);
```

Comment:
Not a big matter, but I'd like to rename the variable `dev` -> `devno` or `devnum`, as `device` is used for `sd_device`, and using `dev` is slightly misleading.